test/metabase/upload_test.clj

```diff
@@ -1022,89 +1026,87 @@
-              (let [new-model (upload-example-csv! :schema-name "not_public" :sync-synchronously? false)
-                    new-table (t2/select-one :model/Table (:table_id new-model))]
-                (is (=? {:display          :table
-                         :database_id      db-id
-                         :dataset_query    {:database db-id
-                                            :query    {:source-table (:id new-table)}
-                                            :type     :query}
-                         :creator_id       (mt/user->id :rasta)
-                         :name             #"(?i)example csv file(.*)"
-                         :collection_id    nil}
-                        new-model)
-                    "A new model is created")
-                (is (=? {:name      #"(?i)example(.*)"
-                         :schema    #"(?i)not_public"
-                         :is_upload true}
-                        new-table)
-                    "A new table is created")
-                (is (= "complete"
-                       (:initial_sync_status new-table))
-                    "The table is synced and marked as complete")
-                (is (= #{["_mb_row_id" :type/PK]
-                         ["id"   :type/PK]
-                         ["name" :type/Name]}
-                       (->> (t2/select Field :table_id (:id new-table))
-                            (map (fn [field] [(u/lower-case-en (:name field))
-                                              (:semantic_type field)]))
-                            set))
-                    "The sync actually runs")
-                (is (true? @in-future?)
-                    "Table has been synced in a separate thread"))))
-          (finally
-            (t2/update! :model/Database db-id original-sync-values)))))))
+    (let [db                   (mt/db)
+          db-id                (u/the-id db)
+          original-sync-values (select-keys db [:is_on_demand :is_full_sync])
+          in-future?           (atom false)
+          schema-name          (or (sql.tx/session-schema driver/*driver*) "public")
+          _                    (t2/update! :model/Database db-id {:is_on_demand false
+                                                                  :is_full_sync false})]
+      (try
+        (with-redefs [ ;; do away with the `future` invocation since we don't want race conditions in a test
+                      future-call (fn [thunk]
+                                    (swap! in-future? (constantly true))
+                                    (thunk))]
+          (testing "Happy path with schema, and without table-prefix"
+            ;; make sure the schema exists
+            (jdbc/execute! (sql-jdbc.conn/connection-details->spec driver/*driver* (:details db))
+                           (format "CREATE SCHEMA IF NOT EXISTS %s;" (sql.tx/qualify-and-quote driver/*driver* schema-name)))
```

Comment:
This is a fairly dense expression, and we repeat it in this file, so I think a helper method is warranted.

Reply:
We don't repeat it in this file. Anyway I've split it into two expressions with a let now, so hopefully it's less dense to read c4e70a700c230f9d7b8779a0c96ba34b35df52e3

Reply:
Definitely looks better. I didn't check whether both cases were in the same file. Avoiding cross test dependencies and keeping tests as self contained as possible makes.

Reply:
I still can't find it in other files, but anyway good point for previous versions of these tests, I should have done that.